lib/puppet/provider/package/dnfmodule.rb

```diff
@@ -0,0 +1,88 @@
+# dnfmodule - A puppet package provider for DNF modules
+#
+# Installing a module:
+#  package { 'postgresql':
+#   provider => 'dnfmodule',
+#   ensure   => '9.6',  # install a specific stream
+#   flavor   => 'client',  # install a specific profile
+# }
+
+
+require 'puppet/provider/package'
+
+Puppet::Type.type(:package).provide :dnfmodule, :parent => :dnf do
+
+  has_feature :installable, :uninstallable, :versionable
+  #has_feature :upgradeable
+  # it's not (yet) feasible to make this upgradeable since module streams don't
+  # always have matching version types (i.e. idm has streams DL1 and client,
+  # other modules have semver streams, others have string streams... we cannot
+  # programatically determine a latest version for ensure => 'latest'
+
+  commands dnf: '/usr/bin/dnf'
+
+  def self.current_version
+    @current_version ||= dnf('--version').split.first
```

Comment:
Can the `dnf --version` command return nil?

Reply:
I don't think so. I've rummaged through their source code and tried it on older Linux distros and it always returns the same format.